### PR TITLE
Add MaxParallelImagePulls support

### PR DESCRIPTION
**Pull request by ruiwen-zhao**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
This PR implements KEP-3673: Kubelet limit of Parallel Image Pulls: https://github.com/kubernetes/enhancements/pull/3713
#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:
This PR:
1. introduces a *int32 type kubelet config MaxParallelImagePulls, which defaults to 0, meaning no limit will be applied to parallel image pulls.
2. validates that MaxParallelImagePulls is non-negative.
3. adds a token channel to parallel puller to enforce the parallel pulls.
4. adds unit test.

#### Does this PR introduce a user-facing change?

```release-note
kubelet: a "maxParallelImagePulls" field can now be specified in the kubelet configuration file to control how many image pulls the kubelet can perform in parallel.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
- [KEP]: KEP-3673 https://github.com/kubernetes/enhancements/pull/3713
- 

```docs

```

### CODE REVIEW

**Comment by bart0sh** [comment]
@ruiwen-zhao please fix failing CI tests, thanks.

File: pkg/kubelet/images/puller.go
Comment (samuelkarp) [inline pkg/kubelet/images/puller.go:43]: If you don't need a value, you can consume less space with a `struct{}` than a `bool`.

```suggestion
	tokens       chan struct{}
```

File: pkg/kubelet/images/image_manager_test.go
Comment (samuelkarp) [inline pkg/kubelet/images/image_manager_test.go:382]: You should be able to drop the sleep here if you reorder:

```suggestion
	wg.Wait()
	fakeRuntime.AssertCallCounts("PullImage", 7)
```

All the goroutines should be complete by the time `wg.Wait()` returns.

**Comment by bart0sh** [comment]
@ruiwen-zhao please fix failing CI tests, thanks

**Comment by pacoxu** [comment]
/assign 
I will review it ASAP.

File: cmd/kubelet/app/options/options.go
Comment (SergeyKanzhelev) [inline cmd/kubelet/app/options/options.go:501]: do we have to add command line arguments? Maybe only have it in config?

File: pkg/kubelet/images/image_manager_test.go
Comment (pacoxu) [inline pkg/kubelet/images/image_manager_test.go:317]: You can try to run the test with stress to make sure there is no flakes.

File: staging/src/k8s.io/kubelet/config/v1beta1/types.go
Comment (pacoxu) [inline staging/src/k8s.io/kubelet/config/v1beta1/types.go:487]:
```
// Setting it to 0 means no limit.
```

File: cmd/kubelet/app/options/options.go
Comment (pacoxu) [inline cmd/kubelet/app/options/options.go:499]: Should we deprecate these two params and suggest users use MaxParallelImagePulls in the future?

File: pkg/kubelet/images/puller.go
Comment (ruiwen-zhao) [reply pkg/kubelet/images/puller.go:43]: good suggestion!

File: cmd/kubelet/app/options/options.go
Comment (ruiwen-zhao) [reply cmd/kubelet/app/options/options.go:499]: I guess you mean RegistryPullQPS and RegistryBurst. 

Yeah, I cannot find the rationale behind why they were added in the first place. We should discourage people from using them and encourage using maxParallelImagePulls once maxParallelImagePulls goes Beta, and then remove these two after maxParallelImagePulls goes GA

File: pkg/kubelet/images/image_manager_test.go
Comment (ruiwen-zhao) [reply pkg/kubelet/images/image_manager_test.go:317]: Can you elaborate on what kind of stress? Or just running the test repeatedly?

File: cmd/kubelet/app/options/options.go
Comment (SergeyKanzhelev) [reply cmd/kubelet/app/options/options.go:501]: command line arguments will be slowly deprecated one by one. I'd suggest to not include it as a command line argument so there will be no need to deprecate it later.

File: cmd/kubelet/app/options/options.go
Comment (SergeyKanzhelev) [reply cmd/kubelet/app/options/options.go:499]: Can you create a bug to deprecate these two? We can use bug to collect feedback.

Comment (pacoxu) [reply cmd/kubelet/app/options/options.go:499]: > Can you create a bug to deprecate these two? We can use bug to collect feedback.

@SergeyKanzhelev  https://github.com/kubernetes/kubernetes/issues/112044 this one was opened before and maybe we can use this one to track it.

File: pkg/kubelet/images/image_manager_test.go
Comment (pacoxu) [reply pkg/kubelet/images/image_manager_test.go:317]: `go install golang.org/x/tools/cmd/stress@latest`
```
   go test -c -race ./pkg/kubelet/images  -run ^TestMaxParallelImagePullLimit$\n
   /Users/pacoxu/go/bin/stress -p 4 ./images.test
 ```
 for instance.

File: cmd/kubelet/app/options/options.go
Comment (ruiwen-zhao) [reply cmd/kubelet/app/options/options.go:501]: gotcha, makes sense. Removed it from command line

File: pkg/kubelet/images/image_manager_test.go
Comment (ruiwen-zhao) [reply pkg/kubelet/images/image_manager_test.go:317]: Uh thanks. Ran the test manually:

```
ruiwen@ruiwen:~/ruiwen-zhao/kubernetes$    /usr/local/google/home/ruiwen/go/bin/stress -p 4 ./images.test
...
3m0s: 224 runs so far, 0 failures
```

File: cmd/kubelet/app/options/options.go
Comment (pacoxu) [inline cmd/kubelet/app/options/options.go:501]:
```suggestion

```

File: staging/src/k8s.io/kubelet/config/v1beta1/types.go
Comment (SergeyKanzhelev) [inline staging/src/k8s.io/kubelet/config/v1beta1/types.go:485]:
```suggestion
	// maxParallelImagePulls sets the maximum number of image pulls in parallel.
```

Comment (ruiwen-zhao) [reply staging/src/k8s.io/kubelet/config/v1beta1/types.go]: Thanks for catching it!

File: pkg/kubelet/images/image_manager_test.go
Comment (ruiwen-zhao) [reply pkg/kubelet/images/image_manager_test.go]: Thanks for catching it!

File: staging/src/k8s.io/kubelet/config/v1beta1/types.go
Comment (SergeyKanzhelev) [inline staging/src/k8s.io/kubelet/config/v1beta1/types.go:490]: I'm surprised that NO similar fields like `MaxPods` fields are being validated to be non-negative numbers. Specifically for `MaxParallelImagePulls` the code will treat negative as `0` as written, so it is not a problem with this PR. Just a general comment

File: staging/src/k8s.io/kubelet/config/v1beta1/types.go
Comment (SergeyKanzhelev) [inline staging/src/k8s.io/kubelet/config/v1beta1/types.go:488]: just thinking out loud, if one wants to run kubelet in a mode when all pods has an image pull policy to be `Never`, would they want this to be `0` indicating "no pulls, please"?

**Review by SergeyKanzhelev — APPROVED** [review]
/lgtm

comments are just thinking out loud

File: staging/src/k8s.io/kubelet/config/v1beta1/types.go
Comment (pacoxu) [reply staging/src/k8s.io/kubelet/config/v1beta1/types.go:488]: If `0` means `no pulls, please`, we should set a negative number for no limit.

We have discussed how this feature goes with promoting parallel pull images.
I supposed it should be something like 
- now(alpha): `SerializeImagePulls=true` and the true limit is 0(for unlimit). 
- later(beta): `SerializeImagePulls=false` and the limit should be 5 by default.(registryPullQPS is by default 5)


If `0` means `no pulls, please`, we should re-think about the default value:
I supposed it should be something like 
- now(alpha): `SerializeImagePulls=true` and the true limit is 1 which is the same meaning as `SerializeImagePulls=true`. 
- later(beta): `SerializeImagePulls=false` and the limit should be 5 by default.(registryPullQPS is by default 5)

**Review by pacoxu — APPROVED** [review]
/lgtm

File: staging/src/k8s.io/kubelet/config/v1beta1/types.go
Comment (pacoxu) [reply staging/src/k8s.io/kubelet/config/v1beta1/types.go:490]: If we want to use 0 as no limit and in the future we use N as default, we should using `int32*` here. Then this can be changed to 0.

If the defaulting is like below using `int32` type, the 0 will be changed to N by default. 0 will not stand for no limit then.
```
if xx.MaxParallelImagePulls == 0 {
	xx.MaxParallelImagePulls = N
}
```

So this should be  `int32*`
```
if xx.MaxParallelImagePulls == nil {
	xx.MaxParallelImagePulls = utilpointer.Int32Ptr(N)
}
```

Comment (ruiwen-zhao) [reply staging/src/k8s.io/kubelet/config/v1beta1/types.go:490]: uh good catch! Will update the code

Comment (ruiwen-zhao) [reply staging/src/k8s.io/kubelet/config/v1beta1/types.go]: Updated the code. PTAL!

**Comment by pacoxu** [comment]
Ping @mrunalp and @dchen1107 for sig-node approval.

> This PR [may require API review](https://git.k8s.io/community/sig-architecture/api-review-process.md#what-apis-need-to-be-reviewed).

@liggitt Do you have time to take a look at the api change here?

File: pkg/kubelet/apis/config/validation/validation.go
Comment (liggitt) [inline pkg/kubelet/apis/config/validation/validation.go:125]: I expected a non-nil MaxParallelImagePulls value to be required to be >= 2 (1 image pull is serial, not parallel)

I also expected `SerializeImagePulls: true` and non-nil MaxParallelImagePulls to be rejected in validation, not to ignore MaxParallelImagePulls

File: staging/src/k8s.io/kubelet/config/v1beta1/types.go
Comment (liggitt) [reply staging/src/k8s.io/kubelet/config/v1beta1/types.go:488]: > Setting it to 0 means no limit.

setting it to null should mean no limit. avoid using a max of 0 when we can simply omit the max field value instead

I think this is how we would specify various options:

no parallel pulls:
```yaml
serializeImagePulls: true # optional, this is the default
maxParallelImagePulls: null
```

parallel pulls with no limit:
```yaml
serializeImagePulls: false
maxParallelImagePulls: null
```

max of 2 images to be pulled in parallel at once:
```yaml
serializeImagePulls: false # optional, should be able to be inferred from non-nil maxParallelImagePulls
maxParallelImagePulls: 2
```

I don't think we would change the default of serializeImagePulls in v1beta1 kubelet config file... we could change it in a future v1 version

File: pkg/kubelet/apis/config/v1beta1/defaults.go
Comment (ruiwen-zhao) [reply pkg/kubelet/apis/config/v1beta1/defaults.go:209]: This would override SerializeImagePulls, if a user
1. specifies `SerializeImagePulls=true` or leaves it to be the default, and
2. specifies `MaxParallelImagePulls=5`.

Having both `SerializeImagePulls=true` and `MaxParallelImagePulls=5` seems to be an invalid input, I prefer adding a validation against this case, instead of silently overriding. WDYT?

File: pkg/kubelet/apis/config/v1beta1/defaults.go
Comment (ruiwen-zhao) [reply pkg/kubelet/apis/config/v1beta1/defaults.go:212]: no particular reason. Currently we are passing `MaxParallelImagePulls` as an integer inside kubelet. I can change it to passing it as a pointer.

File: staging/src/k8s.io/kubelet/config/v1beta1/types.go
Comment (ruiwen-zhao) [reply staging/src/k8s.io/kubelet/config/v1beta1/types.go:488]: makes sense. I will change the validation to what Jordan mentioned above, plus ensuring maxParallelImagePulls >= 2.

File: pkg/kubelet/apis/config/validation/validation.go
Comment (ruiwen-zhao) [reply pkg/kubelet/apis/config/validation/validation.go:125]: makes sense. Will update

**Comment by ruiwen-zhao** [comment]
@liggitt Thanks for the comments! I resolved all your comments except for overriding SerializeImagePulls. PTAL.

File: pkg/kubelet/apis/config/v1beta1/defaults.go
Comment (pacoxu) [reply pkg/kubelet/apis/config/v1beta1/defaults.go]: > Having both `SerializeImagePulls=true` and `MaxParallelImagePulls=5` seems to be an invalid input, I prefer adding a validation against this case, instead of silently overriding. WDYT?

Back to what talked about the future behavior that default `SerializeImagePulls =false & MaxParallelImagePulls=5 or n`, Jordon's suggestion may help in the future case.
- defaulting `MaxParallelImagePulls=5` will override `SerializeImagePulls=true` to `SerializeImagePulls=false`. Or after the upgrade, some kubelet may fail for default value change.

The case details would be
- before the upgrade, the setting is `SerializeImagePulls=true` and default `MaxParallelImagePulls` to nil/1.
- after the upgrade, the setting is `SerializeImagePulls=true` and default `MaxParallelImagePulls` to 5. (This would fail if we add that validation.)

Comment (pacoxu) [reply pkg/kubelet/apis/config/v1beta1/defaults.go]: > can we be smarter about defaulting so that someone doesn't have to specify _both_ `serializeImagePulls: false` and `maxParallelImagePulls: 5`?
> 
> ```
> if obj.MaxParallelImagePulls == nil || *obj.MaxParallelImagePulls < 2 {
>   obj.SerializeImagePulls = utilpointer.BoolPtr(true)
> } else {
>   obj.SerializeImagePulls = utilpointer.BoolPtr(false)
> }
> ```

It seems to be duplicate flags and we may deprecate or remove the `serializeImagePulls` once MaxParallelImagePulls is ready.

Comment (ruiwen-zhao) [reply pkg/kubelet/apis/config/v1beta1/defaults.go]: Yeah I am thinking:
Option 1:
1. For now, keep `SerializeImagePulls=true ` and `MaxParallelImagePulls = nil` as default, and return error if `SerializeImagePulls=true` and `MaxParallelImagePulls` is set. 
2. When we are going to change the default `SerializeImagePulls` to false, we should change the defaults to `SerializeImagePulls=false ` and `MaxParallelImagePulls = N (TBD)` at the same time. 

> after the upgrade, the setting is SerializeImagePulls=true and default MaxParallelImagePulls to 5. (This would fail if we add that validation.)

This should not be the case if we are going to follow option 1.

Option 2 will be what Jordan suggests,  i.e. MaxParallelImagePulls takes higher precedence and overrides SerializeImagePulls. By doing this, when we change the default of SerializeImagePulls, we are essentially changing the default of SerializeImagePulls.

Option 3 will be SerializeImagePulls takes higher precedence, and MaxParallelImagePulls will be ignored if `SerializeImagePulls=true`.

I am leaning towards option 1, mostly to avoid silent overriding and to make things as explicit as possible.

Comment (pacoxu) [reply pkg/kubelet/apis/config/v1beta1/defaults.go]: +1 for Option 2. 
The upgrade problem is my concern in Option 1.

Comment (ruiwen-zhao) [reply pkg/kubelet/apis/config/v1beta1/defaults.go]: Even with option 1, we will not have any upgrade problem if we change the default of MaxParallelImagePulls and SerializeImagePull at the same time

Comment (pacoxu) [reply pkg/kubelet/apis/config/v1beta1/defaults.go]: > The case details would be
> 
> * before the upgrade, the setting is `SerializeImagePulls=true` and default `MaxParallelImagePulls` to nil/1.
> * after the upgrade, the setting is `SerializeImagePulls=true` and default `MaxParallelImagePulls` to 5. (This would fail if we add that validation.)

For defaulting configuration, the upgrade would work as expected with both defaulting values changed. 

But if the user only set `SerializeImagePulls=true` in purpose, after the upgrade, it will be conflicted with `MaxParallelImagePulls=5`.

Comment (ruiwen-zhao) [reply pkg/kubelet/apis/config/v1beta1/defaults.go]: Right, in that case where user explicitly sets SerializeImagePulls to true, option 2 will silently override it, which is not ideal right?

what if we go with option 1, and then when we change to default SerializeImagePulls=false, we will default MaxParallelImagePulls only if SerializeImagePulls is false? By doing this, if a user explicitly sets SerializeImagePulls=true, we will not set the MaxParallel and therefore will not break anything.

Comment (pacoxu) [reply pkg/kubelet/apis/config/v1beta1/defaults.go]: > Right, in that case where user explicitly sets SerializeImagePulls to true, option 2 will silently override it, which is not ideal right?

Option 2 is that `MaxParallelImagePulls` is prior to `SerializeImagePulls`.
 
> what if we go with option 1, and then when we change to default SerializeImagePulls=false, we will default MaxParallelImagePulls only if SerializeImagePulls is false? By doing this, if a user explicitly sets SerializeImagePulls=true, we will not set the MaxParallel and therefore will not break anything.

Option 1 needs to set like `SerializeImagePulls` is prior to `MaxParallelImagePulls`.

Either are OK to me.

Comment (liggitt) [reply pkg/kubelet/apis/config/v1beta1/defaults.go]: > This would override SerializeImagePulls

sorry... my suggestion was implicitly inside a `if SerializeImagePulls == nil` block... I agree we should not override user-specified values in default.

If a user sets `serializeImagePulls: true` and `maxParallelImagePulls: 2`, that should be a validation error

Comment (liggitt) [reply pkg/kubelet/apis/config/v1beta1/defaults.go]: defaults are per-version, so we're only talking about the v1beta1 kubelet config file defaults, which could be independent from defaulting behavior in a future v1 config file version

I still think something like this gives the best user experience, only having to set one field:

```go
if obj.SerializeImagePulls == nil {
  if obj.MaxParallelImagePulls == nil || *obj.MaxParallelImagePulls < 2 {
    obj.SerializeImagePulls = utilpointer.BoolPtr(true)
  } else {
    obj.SerializeImagePulls = utilpointer.BoolPtr(false)
  }
}
```

Comment (ruiwen-zhao) [reply pkg/kubelet/apis/config/v1beta1/defaults.go]: gotcha, updated the code. PTAL!

**Comment by liggitt** [comment]
kubelet config / defaulting / validation lgtm

once the implementation has approval, tag me and I can ack the API change

**Review by pacoxu — APPROVED** [review]
/lgtm

File: pkg/kubelet/apis/config/validation/validation.go
Comment (dchen1107) [inline pkg/kubelet/apis/config/validation/validation.go:125]: Why we need (>=2) constraint here? Shouldn't MaxParallelImagePulls == 1 is a valid input, and effectively change the behavior equal to serializing pulling? In that case we can eventually deprecate both parallelImagePulls and serializeImagePulls, and leave only maxParallelImagePulls?

Comment (ruiwen-zhao) [reply pkg/kubelet/apis/config/validation/validation.go]: good point. I updated the code to
1. allow `MaxParallelImagePulls==1`
2. allow `SerializeImagePulls=true && MaxParallelImagePulls==1`, where a serial image puller will be used

File: pkg/kubelet/apis/config/v1beta1/defaults.go
Comment (liggitt) [inline pkg/kubelet/apis/config/v1beta1/defaults.go:212]: if we allow MaxParallelImagePulls to be 1 now, we need this:

```suggestion
		if obj.MaxParallelImagePulls == nil || *obj.MaxParallelImagePulls < 2 {
```

Comment (ruiwen-zhao) [reply pkg/kubelet/apis/config/v1beta1/defaults.go]: Good catch. Updated the code. Thanks!

**Comment by liggitt** [comment]
API defaulting / validation still lgtm once impl has node approval

**Comment by ruiwen-zhao** [comment]
Rebased again.

**Pull request merged**
